test/utility/thread.c

```diff
@@ -0,0 +1,194 @@
+	conn = opts->conn;
+
+	id = __wt_atomic_fetch_addv64(&opts->next_threadid, 1);
+	testutil_check(conn->open_session(conn, NULL, NULL, &session));
+	testutil_check(
+	    session->open_cursor(session, opts->uri, NULL, "append", &cursor));
+
+	buf[0] = '\2';
+	for (recno = 1; opts->running; ++recno) {
+		if (opts->table_type == TABLE_FIX)
+			cursor->set_value(cursor, buf[0]);
+		else {
+			snprintf(buf, sizeof(buf),
+			    "%" PRIu64 " VALUE ------", recno);
+			cursor->set_value(cursor, buf);
+		}
+		testutil_check(cursor->insert(cursor));
+		if (id == 0) {
+			testutil_check(
+			    cursor->get_key(cursor, &opts->max_inserted_id));
+			if (opts->max_inserted_id >= opts->nrecords)
+				opts->running = false;
+		}
+	}
+
+	return (NULL);
+}
+
+/*
+ * Append into a row store table.
+ */
+void *
+thread_insert_append(void *arg)
+{
+	TEST_OPTS *opts;
+	WT_CONNECTION *conn;
+	WT_SESSION *session;
+	WT_CURSOR *cursor;
+	uint64_t i;
+	char kbuf[64];
+
+	opts = (TEST_OPTS *)arg;
+	conn = opts->conn;
+
+	testutil_check(conn->open_session(conn, NULL, NULL, &session));
+	testutil_check(session->open_cursor(
+	    session, opts->uri, NULL, NULL, &cursor));
+
+	for (i = 0; i < opts->nrecords; ++i) {
+		snprintf(kbuf, sizeof(kbuf), "%010d KEY------", (int)i);
+		cursor->set_key(cursor, kbuf);
+		cursor->set_value(cursor, "========== VALUE =======");
+		testutil_check(cursor->insert(cursor));
+		if (i % 100000 == 0) {
+			printf("insert: %d\r", (int)i);
+			fflush(stdout);
+		}
+	}
+	printf("\n");
+
+	opts->running = false;
+
+	return (NULL);
+}
+
+/*
+ * Append to a table in a "racy" fashion - that is attempt to insert the
+ * same record another thread is likely to also be inserting.
+ */
+void *
+thread_insert_race(void *arg)
+{
+	TEST_OPTS *opts;
+	WT_CONNECTION *conn;
+	WT_SESSION *session;
+	WT_CURSOR *cursor;
+	int ret;
+	uint64_t i, value;
+
+	opts = (TEST_OPTS *)arg;
+	conn = opts->conn;
+
+	testutil_check(conn->open_session(conn, NULL, NULL, &session));
+	testutil_check(session->open_cursor(
+	    session, opts->uri, NULL, NULL, &cursor));
+
+	printf("Running insert thread\n");
+	for (i = 0; i < opts->nrecords; ++i) {
+		testutil_check(
+		    session->begin_transaction(session, "isolation=snapshot"));
+		cursor->set_key(cursor, 1);
+		cursor->search(cursor);
+		cursor->get_value(cursor, &value);
+		cursor->set_key(cursor, 1);
+		cursor->set_value(cursor, value + 1);
+		if ((ret = cursor->update(cursor)) != 0) {
+			if (ret == WT_ROLLBACK) {
+				testutil_check(session->rollback_transaction(
+				    session, NULL));
+				i--;
+				continue;
+			}
+			printf("Error in update: %d\n", ret);
+		}
+		testutil_check(session->commit_transaction(session, NULL));
+		if (i % 10000 == 0) {
+			printf("insert: %d\r", (int)i);
+			fflush(stdout);
+		}
+	}
+	if (1 > 10000)
+		printf("\n");
+
+	opts->running = false;
+
+	return (NULL);
+}
+
+/*
+ * Repeatedly walk backwards through the records in a table.
+ */
+void *
+thread_prev(void *arg)
```

Comment:
I wanted to add a `thread_next` function to go along with this.  But it doesn't appear we use this function anywhere.  So I didn't want to add another function that isn't called.  Maybe we should have a test that both repeatedly walks backwards and another thread repeatedly walks forward through a table.